packages/counters.lua

```diff
@@ -1,64 +1,77 @@
-romans = {
+
+local textcase = SILE.require("packages/textcase").exports
+
+local romans = {
   {1000, "M"},
   {900, "CM"}, {500, "D"}, {400, "CD"}, {100, "C"},
   {90, "XC"}, {50, "L"}, {40, "XL"}, {10, "X"},
   {9, "IX"}, {5, "V"}, {4, "IV"}, {1, "I"}
 }
 
-local function romanize(k)
-  str = ""
-  k = k + 0
+local function num2roman(num)
+  local out = ""
+  num = num + 0
   for _, v in ipairs(romans) do
     val, let = unpack(v)
-    while k >= val do
-      k = k - val
-      str = str..let
+    while num >= val do
+      num = num - val
+      out = out .. let
     end
   end
-  return str
+  return out
 end
 
-local function alpha(n)
+local num2alpha = function (num)
   local out = ""
   local a = string.byte("a")
   repeat
-    n = n - 1
-    out = string.char(n%26 + a) .. out
-    n = (n - n%26)/26
-  until n < 1
+    num = num - 1
+    out = string.char(num % 26 + a) .. out
+    num = (num - num % 26) / 26
+  until num < 1
   return out
 end
 
-local icu
-pcall( function () icu = require("justenoughicu") end)
+local icu = require("justenoughicu")
 
 SILE.formatCounter = function(counter)
-  -- If there is a language-specific formatter, use that.
   local lang = SILE.settings.get("document.language")
-  if SILE.languageSupport.languages[lang] and SILE.languageSupport.languages[lang].counter then
-    local res = SILE.languageSupport.languages[lang].counter(counter)
-    if res then return res end -- allow them to pass.
+  local num2string, num2ordinal
+
+  local LS = SILE.languageSupport.languages[lang]
+  if LS then
+    if LS.formatCounter then
+      local result = LS.formatCounter(counter)
+      if result then return result end
+    end
+    num2alpha = LS.num2alpha and LS.num2alpha or num2alpha
+    num2roman = LS.num2roman and LS.num2roman or num2roman
+    num2string = LS.num2string and LS.num2string or num2string
+    num2ordinal = LS.num2ordinal and LS.num2ordinal or num2ordinal
   end
 
   -- If we have ICU, try that
   if icu then
     local display = counter.display
-
     -- Translate numbering style names which are different in ICU
-    if display == "roman" then
-      display = "romanlow"
-    elseif display == "Roman" then
-      display = "roman"
+    if display == "roman" then display = "romanlow"
+    elseif display == "Roman" then display = "roman"
     end
-
-    local ok, res = pcall(function() return icu.format_number(counter.value, display) end)
-    if ok then return res end
+    local ok, result = pcall(function() return icu.format_number(counter.value, display) end)
+    if ok then return result end
   end
 
-  if (counter.display == "roman") then return romanize(counter.value):lower() end
-  if (counter.display == "Roman") then return romanize(counter.value) end
-  if (counter.display == "alpha") then return alpha(counter.value) end
+  if (counter.display == "string") and num2string then return textcase.lowercase(num2string(counter.value)) end
```

Comment:
How can `num2string` etc. be false?

Reply:
You know what, towers of endless (and inextensible) if-statements feel like an opportunity for a dispatch table. I'm leaning towards putting all these display options into a Lua table and then calling (something like) `SILE.counters.formattingFunction[counter.display](counter.value)`.

If we could do that on a per-language basis (*handwave*) then stuff like "ordinal" doesn't need to be baked in, but could be provided by language support files.

Reply:
`num2string` etc. can be false if the language does not support that format.

I thought about hoisting more of this off on the language support module, but at least my first implementation had a LOT more code that would end up being duplicated between language. Passing the smallest items of work off to per-language support and keeping a lot of the logic in core at least had the advantage of not repeating code and bugs in each language and knowing what to expect.

I'm not saying it couldn't be done better though. Your idea might work.

Reply:
A dispatch table is just a _little_ bit harder in the case of a language not having a particular display method. We can handle it with meta tables I think, but the method I used falls back gracefully in the event that you change the language and the new language doesn't support a method.

Reply:
> `num2string` etc. can be false if the language does not support that format. Those functions are instantiated as nil and then replaced with the per-language functions if they exist.

I'm not sure that's true. First you have:

```lua
num2string = LS.num2string and LS.num2string or num2string
```

which gives either the language-specific version or falls back to a "default" implementation if there isn't one. So `num2string` is guaranteed defined here. But then you test it again...

```lua
if (counter.display == "string") and num2string
```

> A dispatch table is just a little bit harder in the case of a language not having a particular display method.

The way I thought about it is similar to the above. You start with a "default" dispatch table in `packages/counters.lua`:

```
{
"string" = num2string,
"String" = function(n) return textcase.titlecase(num2string(counter.value) end,
...
}

SILE.formatCounter = function(counter)
...
return SILE.counters.formattingRoutines[counter.display](counter.value)
```

and then to get the language-specific variants, you overlay its entries on top of the default ones.

Reply:
Incidentally, I'm sure `X and X or Y` is just another way of saying `X or Y`...

Reply:
> So num2string is guaranteed defined here.

The variable is defined, but it is not guarantied to be a function. It's default value is nil. Hence testing it for truthiness works to see if we have either a language specific or default implementation. For example for roman numerals there is a real default implementation, because those are not language specific, but for strings there is not even a default, languages HAVE to define it.